core/opendaq/reader/src/block_reader_impl.cpp

```diff
@@ -163,27 +196,56 @@ ErrCode BlockReaderImpl::readPacketData()
         }
     }
 
-    if (toRead < remainingSampleCount)
+    info.writtenSampleCount += sampleCountToRead;
+    info.prevSampleIndex += sampleCountToRead;
+    info.remainingSamplesToRead -= sampleCountToRead;
+
+    if (info.writtenSampleCount % blockSize == 0)
     {
-        info.prevSampleIndex += toRead;
+        // how many data packet need to be stepped back
+        const auto rewindSamples = overlappedBlockSize;
+        auto rewindPackets = rewindSamples / packetSampleCount;
```

Comment:
If I'm not mistaken, this assumes that all packets in the queue have the same sample count, which is not necessarily the case. Packets in a connection can have varying sizes.

Reply:
Reworked it